File: heapSort.py
def heapSort(lst):
    #last index of unsorted items
    e = len(lst) - 1
    while e > 0:
        heapify(lst, e)
        #swap root with last unsorted item
        lst[0], lst[e] = lst[e], lst[0]
        e -= 1
    return lst

def heapify(lst, e):
    #last parent node in unsorted items
    p = (e + 1) // 2
    while p >= 0:
        percUp(lst, p, e)
        p -= 1

def percUp(lst, p, e):
    #left and right child node index
    l = (p * 2) + 1
    r = (p * 2) + 2
    if l <= e and lst[l] > lst[p]:
        if r <= e and lst[r] > lst[l]:
            lst[r], lst[p] = lst[p], lst[r]
        else:
            lst[l], lst[p] = lst[p], lst[l]
    elif r <= e and lst[r] > lst[p]:
        lst[r], lst[p] = lst[p], lst[r]

File: Java/HeapSort/src/HeapSort/HeapSort.java
package HeapSort;

public class HeapSort {
    public static int[] heapSort(int[] arr){
        //last index of unsorted items
        int e = arr.length - 1;
        while (e > 0){
            heapify(arr, e);
            //swap root with last unsorted item
            swap(arr,0,e);
            e--;
        }
        return arr;
    }
    private static void heapify(int[] arr, int e){
        //last parent node in unsorted items
        int p = Math.floorDiv(e + 1, 2);
        while (p >= 0){
            percUp(arr, p, e);
            p--;
        }
    }
    private static void percUp(int[] arr, int p, int e){
        int l = (p * 2) + 1;
        int r = (p * 2) + 2;
        if (l <= e && arr[l] > arr[p]){
            if (r <= e && arr[r] > arr[l]) {
                swap(arr, p, r);
            } else {
                swap(arr, p, l);
            }
        }
        else if (r <= e && arr[r] > arr[p]){
            swap(arr,p,r);
        }
    }
    private static void swap(int[] arr, int x, int y){
        int tmp = arr[x];
        arr[x] = arr[y];
        arr[y] = tmp;
    }
}
